### PR TITLE
Add javaagent<->application context bridge for HttpRouteHolder

### LINKED ISSUE #4319: Context keys from `instrumentation-api` should be bridged
`instrumentation-api` module can be added to an application classpath, e.g. for library instrumentations. This will result in different `ContextKey` instances e.g. `io.opentelemetry.instrumentation.api.servlet.ServerSpanNaming#CONTEXT_KEY`. For them to work correctly with the agent, they probably has to be bridged or something.

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Resolves #4319 -- there are no more context keys worth bridging in both instrumentation-api and instrumentation-api-semconv, this was the last useful one.

### CODE REVIEW

File: instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpRouteHolder.java [hunk 1]
```diff
@@ -151,10 +156,38 @@ private boolean isBetterRoute(String name) {
...
   static String getRoute(Context context) {
-    HttpRouteHolder httpRouteHolder = context.get(CONTEXT_KEY);
+    HttpRouteHolder httpRouteHolder = context.get(KEY);
     return httpRouteHolder == null ? null : httpRouteHolder.route;
   }
 
+  /**
+   * This method is internal and is hence not for public use. Its API is unstable and can change at
+   * any time.
+   */
```
Comment (trask) [inline instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpRouteHolder.java:166]: I don't think these disclaimers are needed (guessing they were originally public before changing access to reflection?)

**Review by trask — APPROVED** [review]
🎉

File: instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpRouteHolder.java (hunk 1 above)
Comment (trask) [reply instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpRouteHolder.java:166]: wait, how is the MethodHandle able to access this? 🤔

Comment (mateuszrzeszutek) [reply instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpRouteHolder.java]: Removed the disclaimer comments.

File: instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+              MethodType.methodType(applicationHttpRouteHolder, int.class, String.class));
+      applicationGetUpdatedBySourceOrder =
+          lookup.findVirtual(
+              applicationHttpRouteHolder,
+              "internalGetUpdatedBySourceOrder",
+              MethodType.methodType(int.class));
+      applicationGetRoute =
+          lookup.findVirtual(
+              applicationHttpRouteHolder, "internalGetRoute", MethodType.methodType(String.class));
+    } catch (Throwable ignored) {
```
Comment (trask) [inline instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java:120]: I think something's not quite right. I added `printStackTrace` here and got:

```
java.lang.IllegalAccessException: no such method: io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpRouteHolder.internalCreate(int,String)HttpRouteHolder/invokeStatic
	at java.base/java.lang.invoke.MemberName.makeAccessException(MemberName.java:972)
	at java.base/java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1117)
	at java.base/java.lang.invoke.MethodHandles$Lookup.resolveOrFail(MethodHandles.java:3649)
	at java.base/java.lang.invoke.MethodHandles$Lookup.findStatic(MethodHandles.java:2588)
	at io.opentelemetry.javaagent.instrumentation.opentelemetryapi.context.InstrumentationApiContextBridging.<clinit>(InstrumentationApiContextBridging.java:91)
```

File: instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            "CONSUMER_PROCESS_KEY")) {
+      ContextKeyBridge<?, ?> spanKeyBridge = spanKeyBridge(spanKeyName);
+      if (spanKeyBridge != null) {
+        bridges.add(spanKeyBridge);
+      }
+    }
+
+    ContextKeyBridge<?, ?> httpRouteHolderBridge = httpRouteHolderBridge();
+    if (httpRouteHolderBridge != null) {
+      bridges.add(httpRouteHolderBridge);
```
Comment (trask) [inline instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java:43]: and if I remove the http route bridge here, the tests still pass

Comment (mateuszrzeszutek) [reply instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java:43]: Right, the test wasn't properly constructed... fixed that.
Thanks!

File: instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java (hunk 2 above)
Comment (mateuszrzeszutek) [reply instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java]: Now it's all clear: the `lookup()` method is not actually magic and it doesn't magically make things work 😂 

Turns out that the test was incomplete, and only verified one particular scenario (our of many possible).

File: instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/OpenTelemetryApiInstrumentationModule.java [hunk 4]
```diff
@@ -24,6 +25,8 @@ public List<TypeInstrumentation> typeInstrumentations() {
         new ContextInstrumentation(),
         new ContextStorageWrappersInstrumentation(),
         new OpenTelemetryInstrumentation(),
-        new SpanInstrumentation());
+        new SpanInstrumentation(),
+        // instrumentation-api specific instrumentation
+        new HttpRouteStateInstrumentation());
```
Comment (mateuszrzeszutek) [inline instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/OpenTelemetryApiInstrumentationModule.java:30]: The solution to this issue turned out to be a bit more complex than expected -- I don't particularly like this, but I had to insert additional instrumentation-api instrumentation into the opentelemetry-api module, cause it uses the `AgentContextWrapper`. It doesn't add any muzzle dependencies though, so it should be completely neutral for applications that just use opentelemetry-api without instrumentation-api.

**Comment by trask** [comment]
I haven't thought this through, so maybe missing something, but wondering if we can instead instrument `HttpRouteHolder.get()` and `HttpRouteHolder.updateRoute()`, we may not need a context key bridge for this, and then the instrumentation could live outside of the opentelemetry-api bridge

**Comment by trask** [comment]
(a bit like #4786)

**Comment by mateuszrzeszutek** [comment]
> I haven't thought this through, so maybe missing something, but wondering if we can instead instrument `HttpRouteHolder.get()` and `HttpRouteHolder.updateRoute()`, we may not need a context key bridge for this, and then the instrumentation could live outside of the opentelemetry-api bridge

I thought that this would be a bit more difficult, since we'd have to bridge `HttpRouteSource` and `HttpRouteBiGetter` too, not just the `Context`. I left the context key bridging and extracted a separate instrumentation-api instrumentation (like https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4786); I also moved all the instrumentation-api context bridging tests there.

**Review by trask — APPROVED** [review]
thx, I have a couple more thoughts about bridging, will explore in a follow-up 👍

**Pull request merged**